### PR TITLE
terraform: access list support

### PR DESCRIPTION
**Pull request by hugoShaka**
This PR adds access list support to the Terraform provider. [Since we stopped using gogoproto, generating new resources became increasingly complex.](https://github.com/gravitational/teleport/issues/31919)

The access list terraform schema was generated from the protobuf types and not the ones in `api/types`, which causes all kinds of small oddities from a user PoV, and made this PR way more complex.

User-facing limitations:
- the `metadata` field is nested under the `header` one because of the protobuf structure
- the traits is not a map but a list of structs, each struct with a key and a value attribute

Non user-facing limitations the PR had to work around:
- the protobuf time is different from go's time, I initially wanted to create `ProtoTimeType`/`ProtoTimeValue` but we cannot embbed a protobuf timestamp directly (because it contains a lock, the terraform
  value must be a reference, which is not supported by the protoc generator). The workaround is to use custom types (`Timestamp` and `Duration`). Requires: https://github.com/gravitational/protoc-gen-terraform/pull/36
- the provider has to do the conversion between the proto type and the api/type. This required writing a new provider template for new non-gogo resources (`gen/plural_data_source_new.go.tpl` and `gen/plural_resources_new.go.tpl`)
- The time type difference caused conversion issues fixed by https://github.com/gravitational/teleport/pull/32135

### CODE REVIEW

**Comment by hugoShaka** [comment]
Note: linting and tests are all over the place because I changed the `teleport` and `teleport/api` references to `origin/master`. Once https://github.com/gravitational/teleport/pull/32135 will be merged and backported I'll update the `replace` statements and things should be less red.

File: terraform/gen/plural_data_source_new.go.tpl [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+	}
+
+	{{.VarName}}I, err := r.p.Client.{{.GetMethod}}({{if not .GetWithoutContext}}ctx, {{end}}id.Value{{if ne .WithSecrets ""}}, {{.WithSecrets}}{{end}})
+	if err != nil {
+		resp.Diagnostics.Append(diagFromWrappedErr("Error reading {{.Name}}", trace.Wrap(err), "{{.Kind}}"))
+		return
+	}
+
+    var state types.Object
+	{{.VarName}} := convert.ToProto({{.VarName}}I)
```
Comment (mdwn) [inline terraform/gen/plural_data_source_new.go.tpl:69]: We don't necessarily have to address this now, but it's very possible somebody could stray from this `ToProto`/`FromProto` path in the future. I think it's fine for now, though.

**Review by mdwn — APPROVED** [review]
Overall this LGTM. The new templates look reasonable, but it's a little tough for me to review the very Terraform specific generator code.

File: terraform/gen/plural_data_source_new.go.tpl (hunk 1 above)
Comment (hugoShaka) [reply terraform/gen/plural_data_source_new.go.tpl:69]: I guess we'll have to add a new configuration knob. I hope the generation process gets revamped before we introduce other resources.

File: terraform/gen/main.go [hunk 2]
```diff
@@ -121,7 +123,9 @@ func (p *payload) CheckAndSetDefaults() error {
 
 const (
 	pluralResource          = "plural_resource.go.tpl"
+	pluralResourceNew       = "plural_resource_new.go.tpl"
 	pluralDataSource        = "plural_data_source.go.tpl"
+	pluralDataSourceNew     = "plural_data_source_new.go.tpl"
```
[the remark this answers is not in the document]
Comment (hugoShaka) [reply terraform/gen/main.go:128]: I'm afraid renaming the existing ones and reusing the name will make the file history very confusing.

File: terraform/reference.mdx
[the remark this answers is not in the document]
Comment (hugoShaka) [reply terraform/reference.mdx:1]: what's the process? manual copy in a separate PR?

[the remark this answers is not in the document]
Comment (marcoandredinis) [reply terraform/reference.mdx:1]: Yeah
Open the PR in `master` and `branch/v<latest>` with the new reference.mdx

File: terraform/protoc-gen-terraform-accesslist.yaml [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+# These import paths were not being automatically picked up by
+# protoc-gen-terraform without these overrides
+import_path_overrides:
+    "types": "github.com/gravitational/teleport/api/types"
+    "wrappers": "github.com/gravitational/teleport/api/types/wrappers"
+    "durationpb": "google.golang.org/protobuf/types/known/durationpb"
+    "timestamppb": "google.golang.org/protobuf/types/known/timestamppb"
+    "v1": "github.com/gravitational/teleport/api/gen/proto/go/teleport/header/v1"
+    "v11": "github.com/gravitational/teleport/api/gen/proto/go/teleport/trait/v1"
```
[the remark this answers is not in the document]
Comment (hugoShaka) [reply terraform/protoc-gen-terraform-accesslist.yaml:19]: I tried, but the name is automatically detected `protoc-gen-terraform`. We can add custom imports in the plugin config, but doesn't change the package name used by the generated code, only in the imports.

The generated import end up like
```go
import (
	v1 "v1"
	v11 "v11"

	proto "github.com/gogo/protobuf/proto"
	github_com_gravitational_teleport_api_gen_proto_go_teleport_accesslist_v1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/accesslist/v1"
	_ "github.com/gravitational/teleport/api/gen/proto/go/teleport/header/v1"
	_ "github.com/gravitational/teleport/api/gen/proto/go/teleport/trait/v1"
)
```

**Comment by hugoShaka** [comment]
I hacked a bit, a breaking change in the plugin lib was pushed to `branch/v14` and the plugins repo were not updated to deal with the break. I created a ref (`hugo/aaaaaa`) just before the breaking commit, until we fix the plugins we must not update the teleport and teleport/api versions.

File: terraform/protoc-gen-terraform-accesslist.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (marcoandredinis) [reply terraform/protoc-gen-terraform-accesslist.yaml:19]: That's ok :+1:

File: terraform/gen/main.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (marcoandredinis) [reply terraform/gen/main.go:128]: Can we just not name them "new"?
I have nightmares with file names named `Report new edited.doc.final.__new (3).docx`

Maybe `without_gogo` or `_buf` ?

**Comment by marcoandredinis** [comment]
> a breaking change in the plugin lib was pushed to branch/v14 and the plugins repo were not updated to deal with the break

What change was it?

**Review by mdwn — APPROVED** [review]

**Review by Joerger — APPROVED** [review]

**Pull request merged**